Sources/AppcuesKit/Push/UNUserNotificationCenter+AutoConfig.swift

```diff
@@ -0,0 +1,180 @@
+    ) {
+        // Behavior for all Appcues notification
+        if #available(iOS 14.0, *) {
+            completionHandler([.banner, .list])
+        } else {
+            completionHandler(.alert)
+        }
+    }
+}
+
+extension AppcuesUNUserNotificationCenterDelegate {
+    class WeakPushMonitoring {
+        weak var value: PushMonitoring?
+
+        init (_ wrapping: PushMonitoring) { self.value = wrapping }
+    }
+}
+
+extension UNUserNotificationCenter {
+
+    static func swizzleNotificationCenterGetDelegate() {
+        // this will swap in a new getter for UNUserNotificationCenter.delegate - giving our code a chance to hook in
+        let originalScrollViewDelegateSelector = #selector(getter: self.delegate)
+        let swizzledScrollViewDelegateSelector = #selector(appcues__getNotificationCenterDelegate)
+
+        guard let originalScrollViewMethod = class_getInstanceMethod(self, originalScrollViewDelegateSelector),
+              let swizzledScrollViewMethod = class_getInstanceMethod(self, swizzledScrollViewDelegateSelector) else {
+            return
+        }
+
+        method_exchangeImplementations(originalScrollViewMethod, swizzledScrollViewMethod)
+    }
+
+    // this is our custom getter logic for the UNUserNotificationCenter.delegate
+    @objc
+    private func appcues__getNotificationCenterDelegate() -> UNUserNotificationCenterDelegate? {
+        let delegate: UNUserNotificationCenterDelegate
+
+        var shouldSetDelegate = false
+
+        // this call looks recursive, but it is not, it is calling the swapped implementation
+        // to get the actual delegate value that has been assigned, if any - can be nil
+        if let existingDelegate = appcues__getNotificationCenterDelegate() {
+            delegate = existingDelegate
+        } else {
+            // if it is nil, then we assign our own delegate implementation so there is
+            // something hooked in to listen to scroll
+            delegate = AppcuesUNUserNotificationCenterDelegate.shared
+            shouldSetDelegate = true
+        }
+
+        Swizzler.swizzle(
+            targetInstance: delegate,
+            targetSelector: NSSelectorFromString("userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:"),
+            replacementOwner: UNUserNotificationCenter.self,
+            placeholderSelector: #selector(appcues__placeholderUserNotificationCenterDidReceive),
+            swizzleSelector: #selector(appcues__userNotificationCenterDidReceive)
+        )
+
+        Swizzler.swizzle(
+            targetInstance: delegate,
+            targetSelector: NSSelectorFromString("userNotificationCenter:willPresentNotification:withCompletionHandler:"),
+            replacementOwner: UNUserNotificationCenter.self,
+            placeholderSelector: #selector(appcues__placeholderUserNotificationCenterWillPresent),
+            swizzleSelector: #selector(appcues__userNotificationCenterWillPresent)
+        )
```

Comment:
@iujames the [scrollview implementation](https://github.com/appcues/appcues-ios-sdk/blob/96dd818342a5eff6ef7099e2c843cc113a6dff21/Sources/AppcuesKit/Presentation/Extensions/UIScrollView%2BScrollObserver.swift#L112-L131) did this where the methods are swizzled every time the getter is called.

This seems like overkill, and maybe the methods could be swizzled once after the delegate getter is swizzled? Not sure if you have any insight here, or I can just try it and see what happens.

Reply:
yeah it does seem like it shouldn't be needed each time, but I'm not 100% sure. I think it was challenging to know where to keep that state perhaps? but maybe some associated object type of state reference would work here?